pkg/phlaredb/block_querier.go

```diff
@@ -949,6 +987,67 @@ func (b *singleBlockQuerier) SelectMatchingProfiles(ctx context.Context, params
 	return iter.NewMergeIterator(maxBlockProfile, false, iters...), nil
 }
 
+func (b *singleBlockQuerier) LabelNames(ctx context.Context, params *typesv1.LabelNamesRequest) ([]string, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "LabelNames Block")
+	defer sp.Finish()
+
+	err := b.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	selectors, err := parseSelectors(params.Matchers)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := b.index.LabelNames()
+	if err != nil {
+		return nil, err
+	}
+
+	if selectors.matchesAll() {
+		return names, nil
+	}
+
+	// Clear the names slice. We need to run the matchers to decide which
+	// names to keep. We want to keep this slice capacity, as there will
+	// never be more names than what index.LabelNames() returns.
+	names = names[:0]
+
+	for _, matchers := range selectors {
+		iter, err := PostingsForMatchers(b.index, nil, matchers...)
+		if err != nil {
+			return nil, err
+		}
+
+		for iter.Next() {
+			labelNames, err := b.index.LabelNamesFor(iter.At())
+			if err != nil {
+				if err == storage.ErrNotFound {
+					continue
+				}
+				return nil, err
+			}
+
+			// Select all the label names which we don't already have.
+			for _, name := range labelNames {
+				index := sort.SearchStrings(names, name)
+				switch {
+				case index == len(names): // Add name to the end of the slice.
+					names = append(names, name)
+				case names[index] != name: // Insert name into the middle of the slice.
+					names = append(names, name)
+					copy(names[index+1:], names[index:])
+					names[index] = name
+				}
+			}
+		}
+	}
```

Comment:
My thinking was:

1. Get all the label names for this index
2. If the selector was "matches all", then we can simply return all the label names
3. If there are matchers, then clear the label names slice, but keep the allocated memory the labels
4. For every posting that satisfies the matchers, find the label names for that posting
5. Select only label names we haven't seen yet